Demo LF programs with latency sensitive LLM inferences

Deeksha-20-99 opened the pull request:
This demo showcases time-sensitive Lingua Franca demos that include two LLM agents and measure both logical and physical time per user query for LLM-based inference. In this demo, we also present the federated execution of time-sensitive LLM-based inference.
- The Judge reactor accepts user input as a physical action, which serves as a quiz question for the two LLM agents.
- The LlmA and LlmB reactors receive the query from the Judge and each invokes its respective LLM model to generate an answer.
- Both agents run concurrently. The LLM that responds first is declared the winner, and its answer is reported along with the logical and physical execution times.

For this demo, the LLM models we have utilized are LLAMA chat models having a billion parameters, which are then quantized. These models are used for inference with the training knowledge, and no context is given.

We also explore federated execution, where each LLM agent could run as a separate federate, coordinated by the RTI. This follows the same principle of the game, except that we try to offload models on different devices (to support the models with higher parameters to be run on devices with greater GPU support).
This demo explores time-sensitive events and the distributed execution of LLM-based systems using Lingua Franca.

Please see this [README](https://github.com/lf-lang/lf-demos/blob/d3d416558569faa8287e1f94e5937e843c5516b6/llm/README.md) for more information.

Deeksha-20-99 committed:
Add README for LF LLM demo

Deeksha-20-99 committed:
Adding work in progress code files for an llm example. Files: llm.py, which calls the llama-2-7b-chat model for simple question and answer, agent_llm.lf, which takes in the user input calls llm agent 1 and llm agent 2.

Deeksha-20-99 committed:
changed the file name of the file to be included in agent_llm.lf

Deeksha-20-99 committed:
Added a quiz game. It is a game between two LLM models answering user questions and the model to respond the fastest wins

Deeksha-20-99 committed:
Updated the  README.md for instructions to run the quiz game

Deeksha-20-99 committed:
Removing the older version of the file agent_llm.lf

Deeksha-20-99 committed:
Modified comments to the program

Deeksha-20-99 committed:
created the files for quiz game between two llm models using main reactor and also added a federated execution

Deeksha-20-99 committed:
Adding the git ignore file

Deeksha-20-99 committed:
Fixed the issue for the judge federate to receive the signal that model is loaded

Deeksha-20-99 committed:
Added the version of files for running on different devices

Deeksha-20-99 committed:
Adding a python script for llama 3.2 1B for jetson orin

Deeksha-20-99 committed:
commented the code for testing

Deeksha-20-99 committed:
Testing Jetson

Deeksha-20-99 committed:
Changed the file names in base class

Deeksha-20-99 committed:
Changed the RTI to jetson

Deeksha-20-99 committed:
corrected the ip for jetson orin

hokeun commented on llm/README.md:
```diff
@@ -0,0 +1,95 @@
+
+# Pre-requisites 
+
+You need Python installed, as llm.py is written in Python.
```
We need Python version information here (e.g., minimum version requirement).

hokeun committed:
Add requirements.txt

hokeun committed:
Move requirements.txt to top dir

Deeksha-20-99 committed:
Adding the organized folders and README.md

Deeksha-20-99 committed:
Updated the correct links for federated_execution and requirements in  README.md

Deeksha-20-99 committed:
Updated the requirements.txt for README.md

Deeksha-20-99 committed:
changed the llm_b import statement

hokeun committed:
Rename directories and remove unnecessary files

Deeksha-20-99 committed:
Added more instruction on how to execute this demo README.md

Deeksha-20-99 committed:
changed the path file names for the python files

Deeksha-20-99 committed:
Added the images folder for README.md

Deeksha-20-99 committed:
Updated the image position on the README.md

hokeun committed:
Revise README for LLM Demo overview and structure

Updated the README to provide a clearer overview and directory structure.

Deeksha-20-99 committed:
corrected the spelling of environment README.md

Deeksha-20-99 committed:
corrected the spelling  README.md

Deeksha-20-99 committed:
Changed the comments and removed the Hugging face token and it will be taken from the CLI login

Deeksha-20-99 committed:
Updated the  README.md for federated execution

Deeksha-20-99 committed:
Corrected the path of the python files

Deeksha-20-99 committed:
Corrected the paths of the images in the README.md

Deeksha-20-99 committed:
added the contributors name README.md

Deeksha-20-99 committed:
Merge branch 'llm' of github.com:lf-lang/lf-demos into llm

Deeksha-20-99 committed:
Removed torch and torchvision since they are dependent on the device

Deeksha-20-99 committed:
corrected few things on the README regarding the different reactors

Deeksha-20-99 committed:
Updated the required python version in the README.md

Deeksha-20-99 committed:
Added a command to check if requirements are installed README.md

Deeksha-20-99 committed:
added the common environment name README.md

Deeksha-20-99 committed:
This is the .py file for running the llm_quiz on laptops

Deeksha-20-99 committed:
Added changed to remove DNET warnings

Deeksha-20-99 committed:
Adding a newline for the code

Deeksha-20-99 committed:
fixed the warnings for llm_quiz_game

Deeksha-20-99 committed:
re-organized the code to remove dependancy of llm.py files

Deeksha-20-99 committed:
commented the code to files

Deeksha-20-99 committed:
Revised hardware and software requirements

Updated hardware and software requirements in README.

Deeksha-20-99 committed:
Revise hardware and software requirements in README for federated execution

Updated hardware and software requirements in README.

Deeksha-20-99 committed:
Clean up formatting in README.md

Removed extra newlines in the README instructions.

Deeksha-20-99 committed:
Revise README with updated model execution output

Updated execution log details and model loading messages.

Deeksha-20-99 committed:
Update README for `torch_dtype` deprecation

Updated README to reflect deprecation of `torch_dtype`.

Deeksha-20-99 committed:
Revise example output in README for quiz functionality

Updated example output in README to reflect new quiz question and answers.

Deeksha-20-99 committed:
Update .gitignore to include __pycache__

hokeun committed:
Merge branch 'main' of github.com:lf-lang/lf-demos into llm

# Conflicts:
#	.gitignore

Deeksha-20-99 committed:
added the llma and llmb reactions in a common one to avoid printing only llma always, based on physical time

Deeksha-20-99 committed:
changed 60 sec to 2 sec

Deeksha-20-99 committed:
fine tuned the base class for correcting the error

Deeksha-20-99 committed:
corrected federated version similar and testing on devices

Deeksha-20-99 committed:
corrected the manual timeout

hokeun committed:
Change RTI address to localhost

Deeksha-20-99 committed:
Added comments, removed the self.terminate and added request_stop()

Deeksha-20-99 committed:
Added a Dr. in the comment

Deeksha-20-99 committed:
Added comments for the reactors removed terminate, added request_stop()

hokeun committed:
Update author information in documentation

Deeksha-20-99 committed:
Added new line and removed Dr.

Deeksha-20-99 committed:
Removed the agents directory

Deeksha-20-99 committed:
Revise timeout and software requirements in README and removed extra new lines.

Deeksha-20-99 committed:
Refine README content and removed the new lines

Deeksha-20-99 committed:
Fix formatting and update execution example in README

hokeun reviewed (approved):
Looks great to me!

edwardalee reviewed (approved):
Looks great!  Two questions:
1. Why print both logical and physical times? Won't these always be identical?
2. Why are there two README files? Are they identical? They seem very similar.

Deeksha-20-99 committed:
Fix logical time reference and update inference times

Updated README to correct logical time reference and adjust inference times.

Deeksha-20-99 committed:
Removed the logical time since it is identical to physical time

Deeksha-20-99 committed:
Revise README for clarity and removed redundant information

Deeksha-20-99 committed:
removed the logical time and fixed the prompt

Deeksha-20-99 committed:
removed the logical time comment

Deeksha-20-99 commented:
Hello Professor Lee (@edwardalee),

Thank you so much for your feedback, Professor. 
1. Yes, the logical and physical times are identical in this setup. Therefore, I removed the logical time and now report only the physical time.
2. I kept a separate README for the federated version for clarity. Since the two READMEs were too similar, I removed the  repetitive setup instructions and retained only the information required for execution of the federated version of the program.

Please let me know if there is any other feedback, Professor.

Best regards,
Deeksha

hokeun committed:
Clarify README reference for installation instructions

Updated the README to clarify the reference to the main README for installation instructions.

hokeun committed:
Clarify printed times in README

Updated the README to specify that only physical times are printed.
